crates/bindings/src/rng.rs

```diff
@@ -0,0 +1,195 @@
+use std::cell::{OnceCell, UnsafeCell};
+use std::fmt;
+use std::marker::PhantomData;
+use std::sync::atomic::{AtomicU64, Ordering::Relaxed};
+
+use rand::distributions::{Distribution, Standard};
+use rand::rngs::StdRng;
+use rand::{RngCore, SeedableRng};
+
+use crate::Timestamp;
+
+scoped_tls::scoped_thread_local! {
+    static RNG: OnceCell<RngContext>
+}
+
+// this ensures that you can't stash an StdbRng somewhere and use it
+// at a later point, breaking atomic reproducibility of transactions.
+static RNG_GENERATION: AtomicU64 = AtomicU64::new(0);
+
+struct RngContext {
+    rng: StdCellRng,
+    generation: u64,
+}
+
+impl RngContext {
+    fn seed() -> Self {
+        Self {
+            rng: StdCellRng::new(StdRng::seed_from_u64(Timestamp::now().micros_since_epoch)),
+            generation: RNG_GENERATION.fetch_add(1, Relaxed),
+        }
+    }
+    fn get_rng(&self) -> StdbRng {
+        StdbRng {
+            generation: self.generation,
+            _marker: PhantomData,
+        }
+    }
+}
+
+pub(crate) fn with_rng_set<R>(f: impl FnOnce() -> R) -> R {
+    RNG.set(&OnceCell::new(), f)
+}
+
+/// Generates a random value.
+///
+/// Similar to [`rand::random()`], but using [`StdbRng`] instead.
+///
+/// See also [`spacetimedb::rng()`][rng()]
+pub fn random<T>() -> T
+where
+    Standard: Distribution<T>,
+{
+    if !RNG.is_set() {
+        panic!("cannot use `spacetimedb::random()` outside of a reducer");
+    }
+    Standard.sample(&mut rng())
+}
+
+/// Retrieve the random number generator for this reducer transaction,
+/// seeded by the timestamp of the reducer call.
+///
+/// If you only need a single random value, use [`spacetimedb::random()`][random].
+///
+/// Can be used in method chaining style, e.g. with [`rand::Rng`]
+/// imported: `spacetimedb::rng().gen_range(0..=10)`. Or, cache it locally
+/// for reuse: `let mut rng = spacetimedb::rng();`.
+///
+/// For more information see [`StdbRng`].
+pub fn rng() -> StdbRng {
+    if !RNG.is_set() {
+        panic!("cannot use `spacetimedb::rng()` outside of a reducer");
+    }
+    RNG.with(|r| r.get_or_init(RngContext::seed).get_rng())
+}
+
+/// A reference to the random number generator for this reducer call.
+///
+/// An instance can be obtained via [`spacetimedb::rng()`][rng()]. Import
+/// [`rand::Rng`] to get access to useful methods
+///
+/// This handle
+/// can only be used in the context of this reducer call; it cannot be
+/// stashed in a global and used in a later reducer call, as that would break
+/// the atomicity of transactions.
+///
```

Comment:
Again, I dispute this point, since the handle doesn't contain any information.

Users shouldn't be stashing anything in globals, but `StdbRng` is not worse than any other type in that regard, and I don't think we should treat it specially.